src/Microsoft.ML.Transforms/WhiteningTransform.cs

```diff
@@ -206,139 +223,240 @@ private static VersionInfo GetVersionInfo()
                 loaderAssemblyName: typeof(WhiteningTransform).Assembly.FullName);
         }
 
-        private readonly ColInfoEx[] _exes;
-
-        private const string RegistrationName = "Whitening";
+        private readonly ColInfo[] _infos;
 
         /// <summary>
         /// Convenience constructor for public facing API.
         /// </summary>
         /// <param name="env">Host Environment.</param>
-        /// <param name="input">Input <see cref="IDataView"/>. This is the output from previous transform or loader.</param>
-        /// <param name="name">Name of the output column.</param>
-        /// <param name="source">Name of the column to be transformed. If this is null '<paramref name="name"/>' will be used.</param>
-        /// <param name="kind">Whitening kind (PCA/ZCA).</param>
-        public WhiteningTransform(IHostEnvironment env,
-            IDataView input,
-            string name,
-            string source = null,
-            WhiteningKind kind = Defaults.Kind)
-                : this(env, new Arguments() { Column = new[] { new Column() { Source = source ?? name, Name = name } }, Kind = kind }, input)
+        /// <param name="inputData">Input <see cref="IDataView"/>. This is the output from previous transform or loader.</param>
+        /// <param name="columns"> TODO </param>
+        internal WhiteningTransform(IHostEnvironment env, IDataView inputData, params ColInfo[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(WhiteningTransform)), GetColumnPairs(columns))
         {
         }
 
         /// <summary>
-        /// Public constructor corresponding to SignatureDataTransform.
+        /// Constructor corresponding to SignatureDataTransform.
         /// </summary>
-        public WhiteningTransform(IHostEnvironment env, Arguments args, IDataView input)
-            : base(env, RegistrationName, Contracts.CheckRef(args, nameof(args)).Column,
-                input, TestColumn)
+        internal WhiteningTransform(IHostEnvironment env, Arguments args, IDataView inputData)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(WhiteningTransform)), GetColumnPairs(args.Column))
```

Comment:
Why?
We usually cast arguments to ColumnInfo class and then call Transfrom(env,cols).MakeDataTransform(input)

Reply:
I think I resolved the issue that you are talking about but let me double check. 

Now I have: 
Transfrom(env,input,cols).MakeDataTransform(input)

I remove the args, but I still think I need the input data? Is this correct?

---
In reply to: [229459092](https://github.com/dotnet/machinelearning/pull/1452#discussion_r229459092) [](ancestors = 229459092)